### PR TITLE
FEATURE: Add support for automatic detection of subtypes from Java 17 sealed types

### PR DESCRIPTION
**Pull request by sigpwned**
This (draft) PR adds support for detecting subtypes for serialization automatically using Java 17 sealed types/permitted subclasses, per [the discussion in jackson-future-ideas](https://github.com/FasterXML/jackson-future-ideas/discussions/81#discussioncomment-12506251).

The implementation is pretty simple. (This is largely due to Jackson’s modular design, so kudos to the team!) Happily, it also turned out to be essentially a direct port from [the module I wrote for this feature for Jackson 2.x versions](https://github.com/sigpwned/jackson-modules-java-17-sealed-classes), which increases my confidence that the implementation is at least heading in the right direction.

Hopefully the changes are pretty easy to grok, but I’ll add a little voiceover here, just in case the code I’ve written isn’t as clear as I hope it is.

## Design

The feature is designed to interpret a sealed class declaration like this:

    public sealed class ExampleParent permits ExampleChildA, ExampleChildB {
    }
    
    public final class ExampleChildA extends ExampleParent {
    }
    
    public final class ExampleChildB extends ExampleParent {
    }

As equivalent to the following Jackson 2.14+ structure:

    @JsonSubTypes({
        @JsonSubTypes.Type(value=ExampleChildA.class),
        @JsonSubTypes.Type(value=ExampleChildB.class)})
    public class ExampleParent {
    }
    
    public class ExampleChildA extends ExampleParent {
    }
    
    public class ExampleChildB extends ExampleParent {
    }

The `@JsonTypeInfo` and `@JsonTypeName` annotations should be honored by the implementation as well, so this code should work like users expect, too:

    @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
    public sealed class ExampleParent permits ExampleChildA, ExampleChildB {
    }
    
    @JsonTypeName(“a”)
    public final class ExampleChildA extends ExampleParent {
    }
    
    @JsonTypeName(“b”)
    public final class ExampleChildB extends ExampleParent {
    }

That is: (a) serialization should store type names in serialized objects as a property named `type`; and (b) the `ExampleChildA` and `ExampleChildB` classes should use names `"a"` `"b"`, respectively.

I chose this design because removing redundant `@JsonSubTypes` declarations was what originally drove me to explore this feature as a module back in 2022. However, I'm very happy to revisit if desired!

To @pjfanning’s point, this feature should provide a clearer, cleaner, simpler path towards polymorphism for closed type hierarchies. To @cowtowncoder’s point, I think this approach is also quite ergonomic. I particularly like how DRY it is!

## Implementation

Since Jackson 3.x requires Java 17+, my strategy was to treat subtype detection from sealed types as a first-class citizen. Therefore, I added the feature directly into the `JacksonAnnotationIntrospector` class, as opposed to putting it into a module or something. The `AnnotationIntrospector#findSubtypes` hook made this very easy.

Also, I used Java 17 syntax in places where I thought it made the code clearer. Obviously, this does not match the style of code around it, since the 3.x branch is the first one to allow Java 17. I will happily match style if preferred, just let me know.

## Testing

For testing, my goal was to identify the “core” unit tests for subtype detection using the `@JsonSubTypes` annotation, and modify those tests for sealed class subtype detection. I elected to use `tools.jackson.databind.jsontype.TestSubtypes` for this. I will happily add more tests as needed, just point me in the right direction.

## Open Questions

I have opened a draft PR specifically to address the following open questions:

* I have not touched documentation, but would be happy to. If I can be of use for that, just point me in the right direction.
* Are there any other features of the subtype system that require additional support and/or tests? Based on my read of `JacksonAnnotationIntrospector#findSubtypes` and its uses, I think the implementation is complete, but obviously you guys are the experts!
* In the existing `TestSubtyptes` tests that I adapted for this feature, the code seemed to register subtypes using `JsonMapper.Builder#registerSubtypes`, sometimes redundantly with `@JsonSubTypes` annotations in many cases. I commented those (redundant) calls out because I wanted to ensure that the JacksonAnnotationIntrospector is registering the subtypes itself from the sealed types. The tests pass, but please review, and let me know if I missed the point.

Thank you all for continuing to maintain this wonderful library! The ease of integration of this new feature shows how well-designed this library continues to be, even as it has grown and changed over the last 10+ years!

### CODE REVIEW

File: src/test/java/tools/jackson/databind/jsontype/TestSubtypesWithSealedTypes.java [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+import tools.jackson.databind.testutil.NoCheckSubTypeValidator;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+/**
+ * Unit tests for detecting sealed types as subtypes. Originally copied from `TestSubtypes`.
+ */
+public class TestSubtypesWithSealedTypes extends DatabindTestUtil
+{
+    @JsonTypeInfo(use=JsonTypeInfo.Id.NAME)
```
Comment (pjfanning) [inline src/test/java/tools/jackson/databind/jsontype/TestSubtypesWithSealedTypes.java:23]: would it be possible to test with more `use=JsonTypeInfo.Id.xxx` values?

Comment (sigpwned) [reply src/test/java/tools/jackson/databind/jsontype/TestSubtypesWithSealedTypes.java:23]: Of course! I'll get that in today.

Comment (sigpwned) [reply src/test/java/tools/jackson/databind/jsontype/TestSubtypesWithSealedTypes.java:23]: I've added some tests, and have ensured that the proposed change has tests for the following values:

* `JsonTypeInfo.Id.CLASS`
* `JsonTypeInfo.Id.MINIMAL_CLASS`
* `JsonTypeInfo.Id.NAME`
* `JsonTypeInfo.Id.SIMPLE_NAME`
* `JsonTypeInfo.Id.DEDUCTION`
* `JsonTypeInfo.As.PROPERTY`
* `JsonTypeInfo.As.WRAPPER_OBJECT`
* `JsonTypeInfo.As.WRAPPER_ARRAY`
* `JsonTypeInfo.As.EXISTING_PROPERTY`

The following values still **do not** have tests:

* `JsonTypeInfo.Id.NONE` -- Based on my read of `NoTypeInfoTest`, this seems to concern how to deserialize interfaces as concrete types on a one-to-one basis, as opposed to subtypes.
* `JsonTypeInfo.Id.CUSTOM` -- Based on my read of `CustomTypeIdResolverTest`, this seems to allow users to manage subtype resolution on their own, as opposed ot using annotations/sealed types/other approaches.
* `JsonTypeInfo.As.EXTERNAL_PROPERTY` -- I was not quickly able to find an existing tests for this value. If you think testing this value is important, then I will take a deeper look!

File: src/main/java/tools/jackson/databind/introspect/JacksonAnnotationIntrospector.java [hunk 2]
```diff
@@ -698,6 +711,15 @@ private List<NamedType> findSubtypesCheckRepeatedNames(String annotatedTypeName,
 
         return result;
     }
+    
+    private List<NamedType> findSubtypesByPermittedSubclasses(MapperConfig<?> config, Annotated a, Class<?> clazz)
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/introspect/JacksonAnnotationIntrospector.java:715]: Add `// @since 2.19` for new/changed methods.

File: src/test/java/tools/jackson/databind/jsontype/ExistingPropertyWithSealedTypesTest.java [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+import com.fasterxml.jackson.annotation.JsonTypeInfo.Id;
+
+import tools.jackson.core.type.TypeReference;
+import tools.jackson.databind.ObjectMapper;
+import tools.jackson.databind.testutil.DatabindTestUtil;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static org.junit.jupiter.api.Assertions.*;
+
+public class ExistingPropertyWithSealedTypesTest extends DatabindTestUtil
```
Comment (cowtowncoder) [inline src/test/java/tools/jackson/databind/jsontype/ExistingPropertyWithSealedTypesTest.java:18]: On naming: let's use common prefix ("SealedTypesWith...") ending with "...Test" for these, to group test classes next to each other.

Older pattern of "TestXxx" is being slowly replaced by "...Test" but let's not add new classes using that (but leave `TestSubtypes` as-is for now, rename can be done with separate PR).

File: src/test/java/tools/jackson/databind/jsontype/TestTypedSerializationWithSealedTypes.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    public class A extends Super {}
+    public class B extends Super {}
+
+    /*
+    /**********************************************************
+    /* Unit tests
+    /**********************************************************
+     */
+
+    private final ObjectMapper MAPPER = new ObjectMapper();
```
Comment (cowtowncoder) [inline src/test/java/tools/jackson/databind/jsontype/TestTypedSerializationWithSealedTypes.java:87]: Newer style; use `newJsonMapper()` helper method.

File: src/test/java/tools/jackson/databind/jsontype/TestTypedSerializationWithSealedTypes.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+     *    being added to Animal entries, because Object.class has no type.
+     *    If type information is included, it will not be useful for deserialization,
+     *    since static type does not carry through (unlike in serialization).
+     *
+     *    But it is not quite clear how type information should be pushed through
+     *    array types...
+     */
+    @SuppressWarnings("unchecked")
+    @Test
+    public void testInArray() throws Exception
```
Comment (cowtowncoder) [inline src/test/java/tools/jackson/databind/jsontype/TestTypedSerializationWithSealedTypes.java:160]: As per comment, not 100% sure how this should work so let's just remove test method from here.

File: src/test/java/tools/jackson/databind/jsontype/TestTypedSerializationWithSealedTypes.java [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        Map<?,?> a2 = (Map<?,?>) l.get(1);
+        assertEquals(3, a2.size());
+        assertEquals(Dog.class.getName(), a2.get(classProp));
+    }
+
+    /**
+     * Simple unit test to verify that serializing "empty" beans is ok
+     */
+    @Test
+    public void testEmptyBean() throws Exception
```
Comment (cowtowncoder) [inline src/test/java/tools/jackson/databind/jsontype/TestTypedSerializationWithSealedTypes.java:195]: Remove from this class (no need to repeat).

File: src/test/java/tools/jackson/databind/jsontype/TestTypedSerializationWithSealedTypes.java [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+    {
+        ObjectMapper m = jsonMapperBuilder()
+                .configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false)
+                .build();
+        assertEquals("{\"@type\":\"empty\"}", m.writeValueAsString(new Empty()));
+    }
+
+    @Test
```
Comment (cowtowncoder) [inline src/test/java/tools/jackson/databind/jsontype/TestTypedSerializationWithSealedTypes.java:203]: Not even sure what it's trying to test, let's remove (I realize this is a copy of existing test, but no need to copy).

**Comment by cowtowncoder** [comment]
So far looks good: implementation as simple as expected. Added a note on test class naming (newer pattern to end class name with "Test", not Start; using name prefix to group together).
Other than that can/should trim away test methods that do not relate to specific variation (like: no need to test serialization of "empty beans" more than once).

Hoping to go over PR with more thought later on but wanted to give some quick feedback despite time crunch on my end.

File: src/main/java/tools/jackson/databind/introspect/JacksonAnnotationIntrospector.java (hunk 2 above)
Comment (pjfanning) [reply src/main/java/tools/jackson/databind/introspect/JacksonAnnotationIntrospector.java:715]: This is targeted at master so it's since 3.0

**Comment by sigpwned** [comment]
All comments should be addressed by newest commits! Will send CLA over later this evening.

File: src/main/java/tools/jackson/databind/introspect/JacksonAnnotationIntrospector.java (hunk 2 above)
Comment (cowtowncoder) [reply src/main/java/tools/jackson/databind/introspect/JacksonAnnotationIntrospector.java:717]: My bad, yes.

**Comment by cowtowncoder** [comment]
@sigpwned If this is ready to move from draft please mark as Ready for Review.

**Comment by sigpwned** [comment]
CLA submitted!

**Comment by cowtowncoder** [comment]
Whops. Merged it accidentally already. Need to update release notes next then.
